source/extensions/clusters/aggregate/cluster_lb.cc

```diff
@@ -0,0 +1,87 @@
+#include "extensions/clusters/aggregate/cluster_lb.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Clusters {
+namespace Aggregate {
+
+Upstream::HostConstSharedPtr
+AggregateClusterLoadBalancer::chooseHost(Upstream::LoadBalancerContext* context) {
+  int first = -1, total_avail = 0, size = clusters_per_priority_.size();
+  int r = random_.random() % 100 + 1; // [1, 100]
+
+  ClusterVector clusters;
+  for (int p = 0; p < size; ++p) {
+    ClusterVector cs = clusters_per_priority_[p];
+    int avail = 0;
+    // calculate availability for each cluster and update the availability of the priority
+    for (const auto& cluster : cs) {
+      int availability = calculateAvailability(cluster);
+      avail = std::min(100, avail + availability);
+    }
+
+    // select the cluster set
+    if (total_avail < r && total_avail + avail >= r) {
+      clusters = cs;
+    }
+
+    // set first available cluster set
+    if (avail != 0 && first == -1) {
+      first = p;
+    }
+
+    total_avail += avail;
+  }
+
+  if (total_avail < 100) {
+    // panic mode
```

Comment:
this seems like inconsistent panic mode handling from the existing lbs for a few reasons: 1) the panic mode threshold is not taken into account and 2) it routes everything to the lowest priority instead of attempting to spread out load to all priorities

The "dump everything into one priority" handling seems closer to how the existing LBs handle a total availability of zero, where P0 is set to 100%. The approach here seems less than ideal, as P0 with 10% availability and P1 with 50% availability will result in P0 receiving 100% of the traffic.

Reply:
I will take the threshold into consideration. 
1) If the total < 100% and total > threshold, normalize the availability
2) If the total < threshold, panic and dump everything into highest priority